### PR TITLE
inference script added and integrated inference testing

### PR DESCRIPTION
**Pull request by yash3886**
Modified Python Utilities paths in order to fetch the right loopdistribution_grpc_pb file.

Added a log file to store the sequences generated from the grpc inference.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp [hunk 1]
```diff
@@ -153,6 +153,7 @@ void custom_loop_distribution::initPipeCommunication(
     errs() << "Features populated END...\n";
     int *out;
     size_t size;
+    // MLRunner->evaluate<int  >();
```
Comment (svkeerthy) [inline llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: Please cleanup all of your commented codes.

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp [hunk 2]
```diff
@@ -296,6 +297,7 @@ bool custom_loop_distribution::runOnFunction(Function &F) {
     MLRunner->setRequest(&request);
     MLRunner->setResponse(&response);
 
+    outfile.open("grpc_out.log", std::ios_base::app);
```
Comment (svkeerthy) [inline llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: Why is this needed? Please revert.

File: model/LoopDistribution/src/inference.py [hunk 3]
```diff
@@ -20,9 +20,9 @@
 
 sys.path.extend(
     [
-        f"{BUILD_DIR}/Python-Utilities",
-        f"{MODEL_DIR}",
-        f"{BUILD_DIR}/MLCompilerBridge/CompilerInterface/",
+        f"{BUILD_DIR}/tools/MLCompilerBridge/Python-Utilities/",
+        # f"{MODEL_DIR}",
+        f"{BUILD_DIR}/../MLCompilerBridge/CompilerInterface/",
```
Comment (svkeerthy) [inline model/LoopDistribution/src/inference.py:25]: make this as {BUILD_DIR}/tools/MLCompilerBridge/CompilerInterface/

File: model/LoopDistribution/src/multiagentEnv.py [hunk 4]
```diff
@@ -57,7 +57,7 @@
 import operator
 
 sys.path.extend([
-    f"{BUILD_DIR}/MLCompilerBridge/MLModelRunner/gRPCModelRunner/Python-Utilities"
+    f"{BUILD_DIR}tools/MLCompilerBridge/Python-Utilities/"
```
Comment (svkeerthy) [inline model/LoopDistribution/src/multiagentEnv.py]: Use "{BUILD_DIR}/tools/MLCompilerBridge/Python-Utilities/"

**Review by svkeerthy — CHANGES_REQUESTED** [review]

File: model/LoopDistribution/src/multiagentEnv.py (hunk 4 above)
Comment (yash3886) [reply model/LoopDistribution/src/multiagentEnv.py]: done

File: model/LoopDistribution/src/inference.py (hunk 3 above)
Comment (yash3886) [reply model/LoopDistribution/src/inference.py:25]: done

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp (hunk 2 above)
Comment (yash3886) [reply llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: Similarly to the pipes dumping the sequences generated while inference to this log file to compare with pipes and onnx.

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp (hunk 1 above)
Comment (yash3886) [reply llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: done

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp [hunk 5]
```diff
@@ -118,15 +115,12 @@ struct LoopDistributionServerPass
       const ::loopdistribution::LoopDistributionRequest *request,
       ::loopdistribution::LoopDistributionResponse *response) override {
 
-    errs() << "came here\n";
     std::string partition = request->partitionpattern();
 
     if (partition == "Exit") {
-      errs() << "server exit requested\n";
-      AOTRunner->requestExit();
+      // AOTRunner->requestExit();  HAVE TO UNCOMMENT LATER
```
Comment (svkeerthy) [inline llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp:121]: Is this line not needed?

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp [hunk 6]
```diff
@@ -217,31 +197,25 @@ struct LoopDistributionServerPass
     }
 
     // Calculate and return costs
-    errs() << "Received partition:" << partition << "\n";
 
     if (partition == "Exit") {
-      errs() << "server exit requested\n";
-      AOTRunner->requestExit();
```
Comment (svkeerthy) [inline llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp:224]: requestExit is not needed?

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp [hunk 7]
```diff
@@ -217,31 +197,25 @@ struct LoopDistributionServerPass
...
-
     int *status_out;
     MLRunner->evaluate<int *>(status_out, size);
     std::string final_status;
     for (int i = 0; i < size; i++) {
       final_status.push_back((char)status_out[i]);
     }
     if (final_status == "Exit") {
-      errs() << "Costs sent and acknowledged\n";
+      MLBRIDGE_DEBUG(errs() << "Costs sent and acknowledged\n");
```
Comment (svkeerthy) [inline llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp]: Please use LLVM_DEBUG here

Comment (svkeerthy) [reply llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp]: And in next three lines as well

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp [hunk 8]
```diff
@@ -277,14 +256,14 @@ bool custom_loop_distribution::runOnFunction(Function &F) {
...
-      // MLRunner->evaluate<int64_t>();
+      MLRunner->evaluate<int64_t>();
       errs() << this->DistributionSeq << "\n";
       outfile << this->DistributionSeq << "\n";
       distributed_seqs.push_back(this->DistributionSeq);
     }
     outfile.close();
-    // errs() << "Code is Commented\n";
-    // exit(0);
+    errs() << "Code is Commented\n";
```
Comment (svkeerthy) [inline llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: Please cleanup the logs.

Comment (svkeerthy) [inline llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: Please cleanup the logs.

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp [hunk 9]
```diff
@@ -277,14 +256,14 @@ bool custom_loop_distribution::runOnFunction(Function &F) {
...
+    exit(0);
```
Comment (svkeerthy) [inline llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: Why is it uncommented?

File: model/LoopDistribution/src/experiment.py [hunk 10]
```diff
@@ -177,13 +168,13 @@ def policy_mapping_fn(agent_id, episode=None, **kwargs):
...
                                 }),
-        # "vectorization_policy": (None, obs_vectorization_node,
-        #                         )
+       
     }
 
     config["multiagent"] = {
         "policies" : policies,
-        "policy_mapping_fn": function(policy_mapping_fn)
+
```
Comment (svkeerthy) [inline model/LoopDistribution/src/experiment.py]: Cleanup new line.

File: model/LoopDistribution/src/inference.py [hunk 11]
```diff
@@ -422,6 +409,7 @@ def getAdvice(self, request, context):
         distribution_agent = "distribution_agent_{}".format(count)
 
     if use_pipe:
+        print("Use pipe")
```
Comment (svkeerthy) [inline model/LoopDistribution/src/inference.py]: Clean up logs.

File: model/LoopDistribution/src/inference.py [hunk 12]
```diff
@@ -17,13 +17,12 @@
 import traceback
 import sys
 
+import time
```
Comment (svkeerthy) [inline model/LoopDistribution/src/inference.py:20]: Are we using time?

File: model/LoopDistribution/src/multiagentEnv.py [hunk 13]
```diff
@@ -638,75 +490,67 @@ def sendResponse(self, data):
...
+ 
 
     def startServer(self,filePath, functionName, loopId, serverAddress):
         optPath = f"{BUILD_DIR}/bin/opt"
         clangPath = f"{BUILD_DIR}/bin/clang"
 
         # edit server, path
         if(self.use_pipe):
-            cmd = optPath + " -LoopDistributionServer -loopID " + loopId + " -funcName " + functionName + " -lc-function " + functionName + " -lc-lID " + loopId + " --ml-config-path " + CONFIG_DIR + " -S " + filePath + " --use-pipe-loop-dist " + " --loop-dist-data-format " +  self.config["data_format"]  + " -o /dev/null"     # f" --ml-config-path {REPO_DIR}/build_all/config " +
+
```
Comment (svkeerthy) [inline model/LoopDistribution/src/multiagentEnv.py]: Remove the empty line

**Review by svkeerthy — CHANGES_REQUESTED** [review]

File: model/LoopDistribution/src/inference.py (hunk 12 above)
Comment (yash3886) [reply model/LoopDistribution/src/inference.py:20]: Yes because we are using sleep before reset_pipes. 

Comment added as FIXME.

File: model/LoopDistribution/src/inference.py (hunk 11 above)
Comment (yash3886) [reply model/LoopDistribution/src/inference.py]: Done

File: model/LoopDistribution/src/multiagentEnv.py (hunk 13 above)
Comment (yash3886) [reply model/LoopDistribution/src/multiagentEnv.py]: done

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp (hunk 2 above)
Comment (yash3886) [reply llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: removed

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp (hunk 9 above)
Comment (yash3886) [reply llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: removed

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp (hunk 8 above)
Comment (yash3886) [reply llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: done

Comment (yash3886) [reply llvm/lib/Transforms/Scalar/IR2Vec-LOF/custom_loop_distribution/custom_loop_distribution.cpp]: done

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp (hunk 5 above)
Comment (sid18996) [reply llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp:121]: gRPC flow is not fixed. Hence we can take this later.

File: model/LoopDistribution/src/experiment.py (hunk 10 above)
Comment (Yakshimist) [reply model/LoopDistribution/src/experiment.py]: done

File: llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp (hunk 7 above)
Comment (Yakshimist) [reply llvm/lib/Transforms/Scalar/IR2Vec-LOF/LoopDistributionServer/LoopDistributionServer.cpp]: done

**Review by svkeerthy — APPROVED** [review]
